File: kernels/mixdq_extension/op/qconv2d.py
import torch
import mixdq_extension._C

def qconv2d(
    input_int,
    weight_int,
    weight_scale,
    input_scale,
    input_zp,
    scale,
    weight_sum_by_input_channels,
    bias0,
    bias=None,
    stride=1,
    padding=0,
):
    dilation = 1
    return mixdq_extension._C.qconv2d_w8_a8_ohalf(
        input_int, weight_int, weight_scale, input_scale, input_zp,
        scale, weight_sum_by_input_channels, bias0, 
        bias, stride, padding, dilation
    )


if __name__ == "__main__":
    import torch.nn.functional as F

    def run_test(n, h, w, c, k, r, s, pad, stride, dilation=1, bias=True):
        dev = torch.device('cuda:0')

        input_int = torch.randint(-3, 3, 
                                  (n, c, h, w), 
                                  dtype=torch.int8, 
                                  device=dev)
        input_int = input_int.to(memory_format=torch.channels_last)

        weight_int = torch.randint(-3, 3, 
                                   (k, c, r, s),
                                   dtype=torch.int8, 
                                   device=dev)
        weight_int = weight_int.to(memory_format=torch.channels_last)

        weight_scale = 0.1 + torch.rand((k,), dtype=torch.float32, device=dev)
        input_scale = torch.tensor(0.123, dtype=torch.float32, device=dev)
        input_zp = torch.tensor(2.345, dtype=torch.float32, device=dev)

        if bias:
            bias_ = torch.rand((k,), device=dev, dtype=torch.float16)
        else:
            bias_ = None

        output = qconv2d(
            input_int, 
            weight_int,
            weight_scale,
            input_scale,
            input_zp,
            weight_scale*input_scale,
            weight_int.float().sum(dim=1, keepdim=True) if pad>0 else None,
            weight_int.float().sum(dim=[1,2,3])*input_zp if pad==0 else None,
            bias_,
            stride,
            pad,)
        
        def get_reference_int_compute():
            accumulator = F.conv2d(
                input_int.to(torch.float32),
                weight_int.to(torch.float32),
                stride=stride,
                padding=pad,
                dilation=dilation
            )
            
            # bias1
            w_ = weight_int.to(torch.float32).sum(dim=1, keepdim=True)
            a_ = torch.broadcast_to(-1*input_zp, (input_int.shape[0], 1, input_int.shape[2], input_int.shape[3]))
            bias0 = F.conv2d(a_, w_, stride=stride, padding=pad, dilation=dilation)
            
            output = (accumulator + bias0)*weight_scale[None, :, None, None] * input_scale
            
            if bias:
                output += bias_[None, :, None, None]
            return output.to(torch.float16)
        
        def get_reference_fp_compute():
            weight_fp = weight_int.to(torch.float32) * weight_scale[:, None, None, None]
            input_fp = (input_int.to(torch.float32) - input_zp) * input_scale
            fp_output = F.conv2d(input_fp, weight_fp,
                stride=stride,
                padding=pad,
                dilation=dilation)
            output = fp_output.to(torch.float16)
            if bias:
                output += bias_[None, :, None, None]
            return output

        reference_int = get_reference_int_compute()
        reference_fp = get_reference_fp_compute()

        torch.testing.assert_close(reference_int, reference_fp, rtol=1e-2, atol=1e-2)
        torch.testing.assert_close(output, reference_int)
    

    run_test(1, 14, 14, 512, 1024, 3, 3, pad=1, stride=1, bias=True)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=1, stride=1, bias=True)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=1, stride=2, bias=True)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=0, stride=1, bias=False)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=0, stride=1, bias=True)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=0, stride=2, bias=True)
    run_test(1, 14, 14, 512, 1024, 3, 3, pad=0, stride=1, bias=False)
    # small alignment test cases
    run_test(1, 7, 7, 4, 320, 3, 3, pad=1, stride=1, bias=True) 
    run_test(1, 7, 7, 4, 320, 3, 3, pad=0, stride=1, bias=True) 
    run_test(1, 7, 7, 4, 320, 3, 3, pad=1, stride=2, bias=True) 
    run_test(1, 7, 7, 4, 320, 3, 3, pad=0, stride=2, bias=True) 
    run_test(1, 7, 7, 320, 4, 3, 3, pad=1, stride=1, bias=True)
    run_test(1, 7, 7, 320, 4, 3, 3, pad=0, stride=1, bias=True)
    run_test(1, 7, 7, 320, 4, 3, 3, pad=1, stride=2, bias=True)
    run_test(1, 7, 7, 320, 4, 3, 3, pad=0, stride=2, bias=True)
    ### dilation has bugs. DOn't use dilation > 1
    # run_test(1, 14, 14, 512, 1024, 3, 3, pad=0, stride=1, dilation=2, bias=False)
    # run_test(1, 14, 14, 512, 1024, 3, 3, pad=1, stride=2, dilation=2, bias=False)
    # run_test(1, 14, 14, 512, 1024, 3, 3, pad=1, stride=2, dilation=2, bias=True)

    

File: kernels/mixdq_extension/csrc/qconv2d/qconv2d.cc
#include <torch/extension.h>
#include <ATen/ATen.h>
#include <pybind11/stl.h>
#include <cuda_runtime.h>
#include <ATen/cuda/CUDAContext.h>
#include <c10/util/ArrayRef.h>
#include <torch/torch.h>
#include <vector>

#include "qconv2d.h"
#include "cutlass_conv2d_kernels.h"

namespace mixdq {

// forward declaration
at::Tensor activation_zero_point_propagate(
    const int N, const int H, const int W, const int C, const int K,
    const int R, const int S, const int P, const int Q, 
    const int padding_h, const int padding_w,
    const int stride_h, const int stride_w, const int dilation_h,
    const int dilation_w,
    const at::Tensor& weight_sum_by_input_channel,
    const at::Tensor& input_zero_point
);

// C++ function
at::Tensor
qconv2d_w8_a8_ohalf(at::Tensor input_int8,
                    at::Tensor weight_int8,
                    at::Tensor weight_scale,
                    at::Tensor input_scale,
                    at::Tensor input_zero_point,
                    at::Tensor scale,
                    at::optional<at::Tensor> weight_sum_by_input_channels,
                    at::optional<at::Tensor> bias0,
                    at::optional<const at::Tensor> bias,
                    at::optional<int> stride_,
                    at::optional<int> padding_,
                    at::optional<int> dilation_)
{
    // Sanity check of tensor device, types, and layout
    TORCH_CHECK(input_int8.device().is_cuda(), "Input should be on GPU.");
    TORCH_CHECK(input_int8.device() == weight_int8.device(), 
                "input and weight_int8 should be on the same device.");
    TORCH_CHECK(input_int8.device() == weight_scale.device(), 
                "input and weight_scale should be on the same device.");
    TORCH_CHECK(input_int8.device() == input_scale.device(), 
                "input and input_scale should be on the same device.");
    TORCH_CHECK(input_int8.device() == input_zero_point.device(), 
                "input and input_zero_point should be on the same device.");
    if (weight_sum_by_input_channels.has_value()) {
        TORCH_CHECK(
        input_int8.device() == weight_sum_by_input_channels.value().device(),
        "input and weight_sum_by_input_channels should be on the same device.");
    }
    if (bias0.has_value()) {
        TORCH_CHECK(
        input_int8.device() == bias0.value().device(),
        "input and bias0 should be on the same device.");
    }
    if (bias.has_value()) {
        TORCH_CHECK(input_int8.device() == bias.value().device(), 
                    "input and bias should be on the same device.");
    }

    TORCH_CHECK(input_int8.scalar_type() == torch::kInt8, 
                "input_int8 should be int8 type");
    TORCH_CHECK(weight_int8.scalar_type() == torch::kInt8, 
                "weight_int8 should be int8 type");
    TORCH_CHECK(weight_scale.scalar_type() == torch::kFloat32, 
                "Currently only support weight_scale with float32 type");
    TORCH_CHECK(input_scale.scalar_type() == torch::kFloat32, 
                "Currently only support input_scale with float32 type");
    TORCH_CHECK(input_zero_point.scalar_type() == torch::kFloat32, 
                "Currently only support input_zero_point with float32 type");
    if (weight_sum_by_input_channels.has_value()) {
        TORCH_CHECK(
        weight_sum_by_input_channels.value().scalar_type() == torch::kFloat32,
        "Currently only support weight_sum_by_input_channels with float32 type");
    }
    if (bias0.has_value()) {
        TORCH_CHECK(
        bias0.value().scalar_type() == torch::kFloat32,
        "Currently only support bias0 with float32 type");
    }
    if (bias.has_value()) {
        TORCH_CHECK(bias.value().scalar_type() == torch::kFloat16, 
                    "Currently only support bias with float16 type");
    }

    at::Tensor input_int_channels_last = input_int8.contiguous(
        /*memory_format=*/at::MemoryFormat::ChannelsLast);

    at::Tensor weight_int_channels_last = weight_int8.contiguous(
        /*memory_format=*/at::MemoryFormat::ChannelsLast);

    int N, H, W, C;
    N = input_int8.size(0);
    C = input_int8.size(1);
    H = input_int8.size(2);
    W = input_int8.size(3);
    int K, R, S;
    K = weight_int8.size(0);
    R = weight_int8.size(2);
    S = weight_int8.size(3);
    int padding =0, stride =1, dilation =1;
    if (padding_.has_value()) { padding = padding_.value();}
    if (stride_.has_value()) { stride = stride_.value();}
    if (dilation_.has_value()) { dilation = dilation_.value();}
    
    int P, Q;
    P = (H + 2 * padding - dilation * (R -1) -1) / stride + 1;
    Q = (W + 2 * padding - dilation * (S -1) -1) / stride + 1;

    TORCH_CHECK(weight_scale.numel() == K, 
    "The size of the weight_scale vector should be equal to output_channels.");
    if (padding == 0) {
        TORCH_CHECK(bias0.value().numel() == K, 
        "The size of bias0 should equal output_channels.");
    }
    else {
        TORCH_CHECK(weight_sum_by_input_channels.value().numel() == K*R*S, 
        "The size of weight_sum_by_input_channels should equal K*R*S.");
    }
    if (bias.has_value()) {
        TORCH_CHECK(bias.value().numel() == K,
        "The size of the bias vector should be equal to output_channels.");
    }

    // bias0
    at::Tensor bias0_;
    if (padding > 0) {
        bias0_ = activation_zero_point_propagate(
            N, H, W, C, K, R, S, P, Q, padding, padding, stride, stride, dilation,
            dilation, weight_sum_by_input_channels.value(), input_zero_point
        );
        // auto input_zp_broadcast = at::broadcast_to(input_zero_point, {N, 1, H, W});
        // bias0 = torch::nn::functional::conv2d(
        //     input_zp_broadcast.to(torch::kFloat32),
        //     weight_sum_by_input_channels,
        //     torch::nn::functional::Conv2dFuncOptions().stride(stride).padding(padding)
        // ).contiguous(/*memory_format=*/at::MemoryFormat::ChannelsLast);
    }
    else {
        bias0_ = bias0.value();
    }

    // create output tensor

    torch::TensorOptions options = torch::TensorOptions()
                                   .dtype(torch::kHalf)
                                   .device(input_int8.device())
                                   .memory_format(at::MemoryFormat::ChannelsLast);

    at::Tensor D = torch::empty({N, K, P, Q}, options);

    cutlass::Status status;
    static const int alignment_A_optimal = 16;// 128/8bit
    static const int alignment_D_optimal = 8; // 128/16bit
    static const int alignment_A_small = 4; 
    static const int alignment_D_small = 4;

    if (C % alignment_A_optimal == 0 && K % alignment_D_optimal == 0) {
        if (padding > 0) {
            status = cutlass_tensorop_f16_i16832conv2dfprop_with_padding_optimal_alignment(
                N, H, W, C, K, R, S, padding, padding, stride, stride, dilation,
                dilation, input_int_channels_last, weight_int_channels_last, bias0_, 
                scale, bias, 
                D
            );
        }
        else {
            status = cutlass_tensorop_f16_i16832conv2dfprop_no_padding_optimal_alignment(
                N, H, W, C, K, R, S, padding, padding, stride, stride, dilation,
                dilation, input_int_channels_last, weight_int_channels_last, bias0_, 
                scale, bias, 
                D
            );
        }
    }
    else if (C % alignment_A_small == 0 && K % alignment_D_small == 0) {
        if (padding > 0) {
            status = cutlass_tensorop_f16_i16832conv2dfprop_with_padding_small_alignment(
                N, H, W, C, K, R, S, padding, padding, stride, stride, dilation,
                dilation, input_int_channels_last, weight_int_channels_last, bias0_, 
                scale, bias, 
                D
            );
        }
        else {
            status = cutlass_tensorop_f16_i16832conv2dfprop_no_padding_small_alignment(
                N, H, W, C, K, R, S, padding, padding, stride, stride, dilation,
                dilation, input_int_channels_last, weight_int_channels_last, bias0_, 
                scale, bias, 
                D
            );
        }
    }
    else {
        TORCH_CHECK(false, 
        "Int8 kernel with input or output alignment not to 4 is not supported.");
    }
    TORCH_CHECK(status == cutlass::Status::kSuccess, "CUTLASS kernel failed");

    return D;
}

void initQuantizedConv2dBindings(py::module m) {
    m.def("qconv2d_w8_a8_ohalf",
    py::overload_cast<const at::Tensor, 
                      const at::Tensor, 
                      const at::Tensor, 
                      const at::Tensor, 
                      const at::Tensor, 
                      const at::Tensor, 
                      at::optional<at::Tensor>,
                      at::optional<at::Tensor>,
                      at::optional<const at::Tensor>,
                      at::optional<int>,
                      at::optional<int>,
                      at::optional<int>
                      >(&qconv2d_w8_a8_ohalf),
        py::arg("input_int8"),
        py::arg("weight_int8"),
        py::arg("weight_scale"),
        py::arg("input_scale"),
        py::arg("input_zero_point"),
        py::arg("scale"),
        py::arg("weight_sum_by_input_channels"),
        py::arg("bias0"),
        py::arg("bias") = nullptr,
        py::arg("stride") = 1,
        py::arg("padding") = 0,
        py::arg("dilation") = 1
        );
}

}   // namespace mixdq
